File: mobile/lib/core/protocol/relay_engine.dart
import 'dart:io';
import 'dart:typed_data';

import '../model/model.dart';
import 'envelope_codec.dart';
import 'frame.dart';
import 'outbound_scheduler.dart';
import 'secure_envelope.dart';

/// Port of `in.meshsetu.protocol.RelayStore` (Kotlin `RelayEngine.kt`).
///
/// An abstract *class* (not `abstract interface class`) so subclasses can
/// inherit the no-op default bodies, matching Kotlin's `interface { fun f()
/// = default }` methods. Extend this (don't `implements` it), or the
/// defaults won't carry over — Dart's `implements` only takes the member
/// signatures, not their bodies.
abstract class RelayStore {
  void persist(MeshEnvelope envelope);
  bool contains(int objectId) => false;
  void enqueue(EncryptedObject value) {}
  List<EncryptedObject> pending(int nowMs) => const [];
  void markAck(int objectId, String peerId) {}
  void markDeferred(EncryptedObject value) {}
}

class RelayMetric {
  const RelayMetric(
    this.kind, {
    this.objectId,
    this.peerId,
    this.value,
    this.detail,
  });

  final String kind;
  final int? objectId;
  final String? peerId;
  final int? value;
  final String? detail;
}

class RelayResult {
  const RelayResult(this.controlFrames, this.metrics);

  final List<Uint8List> controlFrames;
  final List<RelayMetric> metrics;
}

typedef PersistListener = void Function(MeshEnvelope envelope, String peerId);

const int _maxPartialObjects = 128;
const int _partialTimeoutMs = 30000;
const int _nackDelayMs = 2000;

/// Port of `in.meshsetu.protocol.MeshRelayEngine` (Kotlin `RelayEngine.kt`).
///
/// `receive`/`submit` are `async` here because [CryptoEnvelope] is (see
/// secure_envelope.dart doc comment). Not explicitly locked (see the
/// concurrency note in mesh_transport.dart, which owns serialization).
class MeshRelayEngine {
  MeshRelayEngine({
    required this.siteId,
    required this.crypto,
    required this.store,
    required this.clockMs,
    OutboundScheduler? scheduler,
    RecentObjectCache? dedupe,
    PersistListener? onPersist,
  }) : scheduler = scheduler ?? OutboundScheduler(),
       dedupe = dedupe ?? RecentObjectCache(),
       _onPersist = onPersist ?? ((_, __) {}) {
    final now = clockMs();
    for (final value in store.pending(now)) {
      _knownObjects[value.objectId] = value;
      this.scheduler.enqueue(value, now);
    }
  }

  final String siteId;
  final CryptoEnvelope crypto;
  final RelayStore store;
  final int Function() clockMs;
  final OutboundScheduler scheduler;
  final RecentObjectCache dedupe;
  final PersistListener _onPersist;

  final Map<(String, int), ReassemblyBuffer> _partial = {};
  final Map<(String, int), int> _partialCreatedAt = {};
  final Map<(String, int), int> _lastNackAt = {};
  final Map<int, EncryptedObject> _knownObjects = {};
  final Map<(String, int), int> _inFlight = {};
  final Set<int> _acknowledged = {};
  final List<RelayMetric> _metrics = [];
  final List<PersistListener> _listeners = [];

  void addPersistListener(PersistListener listener) => _listeners.add(listener);

  Future<RelayResult> receive(String peerId, Uint8List encodedFrame) async {
    _cleanupPartial(clockMs());

    final MeshFrame frame;
    try {
      frame = FrameCodec.decode(encodedFrame);
    } catch (_) {
      _metrics.add(RelayMetric('invalid_frame', peerId: peerId));
      return RelayResult(const [], _drain());
    }

    if (frame.type == FrameType.custodyAck) {
      final matches =
          frame.payload.length == 8 &&
          ByteData.sublistView(frame.payload).getInt64(0, Endian.big) ==
              frame.objectId;
      if (!matches) {
        _metrics.add(
          RelayMetric('invalid_ack', objectId: frame.objectId, peerId: peerId),
        );
        return RelayResult(const [], _drain());
      }
      if (_inFlight.remove((peerId, frame.objectId)) != null) {
        _acknowledged.add(frame.objectId);
        store.markAck(frame.objectId, peerId);
        _metrics.add(
          RelayMetric('ack', objectId: frame.objectId, peerId: peerId),
        );
      } else {
        _metrics.add(
          RelayMetric(
            'unexpected_ack',
            objectId: frame.objectId,
            peerId: peerId,
          ),
        );
      }
      return RelayResult(const [], _drain());
    }

    if (frame.type == FrameType.nack) {
      if (frame.payload.isEmpty) {
        _metrics.add(
          RelayMetric('invalid_nack', objectId: frame.objectId, peerId: peerId),
        );
      } else {
        final known = _knownObjects[frame.objectId];
        if (known != null) {
          if (!_acknowledged.contains(frame.objectId)) {
            scheduler.enqueue(known, clockMs());
            _metrics.add(
              RelayMetric(
                'nack_retry',
                objectId: frame.objectId,
                peerId: peerId,
              ),
            );
          }
        } else {
          _metrics.add(
            RelayMetric(
              'unknown_nack',
              objectId: frame.objectId,
              peerId: peerId,
            ),
          );
        }
      }
      return RelayResult(const [], _drain());
    }

    if (frame.type == FrameType.hello) {
      _metrics.add(RelayMetric('hello', peerId: peerId));
      return RelayResult(const [], _drain());
    }

    if (frame.type != FrameType.data) {
      return RelayResult(const [], _drain());
    }

    final key = (peerId, frame.objectId);
    var buffer = _partial[key];
    if (buffer == null) {
      if (_partial.length >= _maxPartialObjects) {
        _metrics.add(
          RelayMetric(
            'reassembly_capacity',
            objectId: frame.objectId,
            peerId: peerId,
          ),
        );
        return RelayResult(const [], _drain());
      }
      buffer = ReassemblyBuffer(frame.count);
      _partial[key] = buffer;
      _partialCreatedAt[key] = clockMs();
    }
    try {
      buffer.add(frame);
    } catch (_) {
      _partial.remove(key);
      _partialCreatedAt.remove(key);
      _lastNackAt.remove(key);
      _metrics.add(
        RelayMetric(
          'reassembly_rejected',
          objectId: frame.objectId,
          peerId: peerId,
        ),
      );
      return RelayResult(const [], _drain());
    }
    if (!buffer.complete()) return RelayResult(const [], _drain());
    _partial.remove(key);
    _partialCreatedAt.remove(key);
    _lastNackAt.remove(key);

    final encrypted = EncryptedObject(
      objectId: frame.objectId,
      trafficClass: _trafficClassFor(frame.priority),
      bytes: buffer.join(),
      expiresAtMs: _maxInt,
    );
    final envelope = await crypto.decrypt(encrypted);
    if (envelope == null) {
      _metrics.add(
        RelayMetric('invalid_object', objectId: frame.objectId, peerId: peerId),
      );
      return RelayResult(const [], _drain());
    }
    if (envelope.siteId != siteId) {
      _metrics.add(
        RelayMetric('wrong_site', objectId: envelope.objectId, peerId: peerId),
      );
      return RelayResult(const [], _drain());
    }
    final now = clockMs();
    if (envelope.expiresAtMs <= now) {
      _metrics.add(
        RelayMetric('expired', objectId: envelope.objectId, peerId: peerId),
      );
      return RelayResult(const [], _drain());
    }
    if (store.contains(envelope.objectId) ||
        !dedupe.markIfNew(envelope.objectId, envelope.expiresAtMs, now)) {
      _metrics.add(
        RelayMetric('duplicate', objectId: envelope.objectId, peerId: peerId),
      );
      return RelayResult([_ack(frame.objectId, frame.priority)], _drain());
    }

    try {
      store.persist(envelope);
    } catch (_) {
      _metrics.add(
        RelayMetric(
          'persist_failed',
          objectId: envelope.objectId,
          peerId: peerId,
        ),
      );
      return RelayResult(const [], _drain());
    }
    _onPersist(envelope, peerId);
    for (final listener in _listeners) {
      listener(envelope, peerId);
    }
    _metrics.add(
      RelayMetric(
        'object_complete',
        objectId: envelope.objectId,
        peerId: peerId,
        value: envelope.hopCount,
      ),
    );
    _metrics.add(
      RelayMetric(
        'object_latency_ms',
        objectId: envelope.objectId,
        peerId: peerId,
        value: (now - envelope.createdAtMs).clamp(0, 0x7FFFFFFF),
      ),
    );
    if (envelope.hopCount < envelope.hopLimit) {
      final relayed = await crypto.encrypt(
        envelope.copyWith(hopCount: envelope.hopCount + 1),
      );
      _enqueue(relayed, now);
      _metrics.add(
        RelayMetric(
          'relay_enqueued',
          objectId: envelope.objectId,
          peerId: peerId,
        ),
      );
    }
    return RelayResult([_ack(frame.objectId, frame.priority)], _drain());
  }

  EncryptedObject? nextOutbound({int? nowMs}) =>
      scheduler.next(nowMs ?? clockMs());

  void requeue(EncryptedObject value, {int? nowMs}) =>
      _enqueue(value, nowMs ?? clockMs());

  void markSent(EncryptedObject value, String peerId, {int? nowMs}) {
    _knownObjects[value.objectId] = value;
    _inFlight[(peerId, value.objectId)] = nowMs ?? clockMs();
  }

  void retryExpired({int? nowMs, int timeoutMs = 8000}) {
    final now = nowMs ?? clockMs();
    final due = _inFlight.entries
        .where((e) => now - e.value >= timeoutMs)
        .map((e) => e.key)
        .toList();
    for (final key in due) {
      _inFlight.remove(key);
      if (!_acknowledged.contains(key.$2)) {
        final known = _knownObjects[key.$2];
        if (known != null) _enqueue(known, now);
      }
      _metrics.add(
        RelayMetric('ack_timeout', objectId: key.$2, peerId: key.$1),
      );
    }
  }

  void defer(EncryptedObject value) {
    _knownObjects.remove(value.objectId);
    store.markDeferred(value);
  }

  Uint8List? missing(String peerId, int objectId, int priority) {
    final buffer = _partial[(peerId, objectId)];
    if (buffer == null) return null;
    return FrameCodec.encode(
      MeshFrame(
        type: FrameType.nack,
        priority: priority,
        flags: 0,
        objectId: objectId,
        sequence: 0,
        count: 1,
        payload: buffer.missingBitmap(),
      ),
    );
  }

  List<Uint8List> missingForPeer(
    String peerId, {
    int priority = 3,
    int? nowMs,
  }) {
    final now = nowMs ?? clockMs();
    final results = <Uint8List>[];
    for (final key in _partial.keys.toList()) {
      if (key.$1 != peerId) continue;
      final createdAt = _partialCreatedAt[key] ?? now;
      if (now - createdAt < _nackDelayMs) continue;
      final lastNack = _lastNackAt[key];
      if (lastNack != null && now - lastNack < _nackDelayMs) continue;
      _lastNackAt[key] = now;
      final frame = missing(peerId, key.$2, priority);
      if (frame != null) results.add(frame);
    }
    return results;
  }

  Future<EncryptedObject> submit(MeshEnvelope envelope, {int? nowMs}) async {
    final encrypted = await crypto.encrypt(envelope);
    _enqueue(encrypted, nowMs ?? clockMs());
    return encrypted;
  }

  void _enqueue(EncryptedObject value, int nowMs) {
    _knownObjects[value.objectId] = value;
    store.enqueue(value);
    scheduler.enqueue(value, nowMs);
  }

  void _cleanupPartial(int nowMs) {
    final expired = _partialCreatedAt.entries
        .where((e) => nowMs - e.value >= _partialTimeoutMs)
        .map((e) => e.key)
        .toList();
    for (final key in expired) {
      _partial.remove(key);
      _partialCreatedAt.remove(key);
      _lastNackAt.remove(key);
      _metrics.add(
        RelayMetric('reassembly_timeout', objectId: key.$2, peerId: key.$1),
      );
    }
  }

  List<RelayMetric> drainMetrics() => _drain();

  List<RelayMetric> _drain() {
    final drained = List<RelayMetric>.of(_metrics);
    _metrics.clear();
    return drained;
  }

  static TrafficClass _trafficClassFor(int priority) => switch (priority) {
    1 => TrafficClass.sosStructured,
    2 => TrafficClass.authorityControl,
    3 => TrafficClass.voiceEvidence,
    4 => TrafficClass.roomMessage,
    _ => TrafficClass.telemetry,
  };

  static Uint8List _ack(int objectId, int priority) {
    final payload = ByteData(8)..setInt64(0, objectId, Endian.big);
    return FrameCodec.encode(
      MeshFrame(
        type: FrameType.custodyAck,
        priority: priority,
        flags: 0,
        objectId: objectId,
        sequence: 0,
        count: 1,
        payload: payload.buffer.asUint8List(),
      ),
    );
  }
}

// Matches Kotlin's `Long.MAX_VALUE` sentinel used for reassembly-scratch
// EncryptedObjects that are immediately consumed by crypto.decrypt and never
// scheduled (so their expiry is irrelevant).
const int _maxInt = 0x7FFFFFFFFFFFFFFF;

/// Port of `in.meshsetu.protocol.FileRelayStore` (Kotlin `RelayEngine.kt`).
///
/// Durable outbox/inbox backed by `dart:io` files under [directory], with
/// the same atomic-write-then-rename pattern as the Kotlin source (write to
/// a `.tmp` file, then rename over the target; rename is atomic on POSIX
/// filesystems when source/target share a directory).
class FileRelayStore extends RelayStore {
  FileRelayStore(this.directory) {
    _inbox.createSync(recursive: true);
    _outbox.createSync(recursive: true);
    Directory('${directory.path}/deferred').createSync(recursive: true);
  }

  final Directory directory;
  Directory get _inbox => Directory('${directory.path}/inbox');
  Directory get _outbox => Directory('${directory.path}/outbox');

  @override
  void persist(MeshEnvelope envelope) => _writeAtomically(
    File('${_inbox.path}/${envelope.objectId}.bin'),
    Uint8List.fromList(EnvelopeCodec.encode(envelope)),
  );

  @override
  bool contains(int objectId) =>
      File('${_inbox.path}/$objectId.bin').existsSync();

  @override
  void enqueue(EncryptedObject value) {
    final out = BytesBuilder();
    out.add(_int64(value.objectId));
    out.add(_int32(value.trafficClass.rank));
    out.add(_int64(value.expiresAtMs));
    out.add(_int64(value.createdAtMs));
    out.add(_int32(value.bytes.length));
    out.add(value.bytes);
    _writeAtomically(
      File('${_outbox.path}/${value.objectId}.bin'),
      out.toBytes(),
    );
  }

  @override
  List<EncryptedObject> pending(int nowMs) {
    final files = _outbox.existsSync() ? _outbox.listSync() : const [];
    final results = <EncryptedObject>[];
    for (final entry in files) {
      if (entry is! File) continue;
      try {
        final bytes = entry.readAsBytesSync();
        final input = ByteData.sublistView(Uint8List.fromList(bytes));
        var offset = 0;
        final objectId = input.getInt64(offset, Endian.big);
        offset += 8;
        final rank = input.getInt32(offset, Endian.big);
        offset += 4;
        final expiresAtMs = input.getInt64(offset, Endian.big);
        offset += 8;
        final createdAtMs = input.getInt64(offset, Endian.big);
        offset += 8;
        final length = input.getInt32(offset, Endian.big);
        offset += 4;
        if (length < 1 || length > maxObjectBytes) continue;
        final payload = Uint8List.sublistView(
          Uint8List.fromList(bytes),
          offset,
          offset + length,
        );
        final trafficClass = TrafficClass.values.firstWhere(
          (t) => t.rank == rank,
        );
        final value = EncryptedObject(
          objectId: objectId,
          trafficClass: trafficClass,
          bytes: payload,
          expiresAtMs: expiresAtMs,
          createdAtMs: createdAtMs,
        );
        if (value.expiresAtMs > nowMs) {
          results.add(value);
        } else {
          entry.deleteSync();
        }
      } catch (_) {
        // Preserve unreadable entries for recovery rather than risking data
        // loss on a transient filesystem failure.
        continue;
      }
    }
    return results;
  }

  @override
  void markAck(int objectId, String peerId) {
    final file = File('${_outbox.path}/$objectId.bin');
    if (file.existsSync()) file.deleteSync();
  }

  @override
  void markDeferred(EncryptedObject value) {
    final outboxFile = File('${_outbox.path}/${value.objectId}.bin');
    if (outboxFile.existsSync()) outboxFile.deleteSync();
    _writeAtomically(
      File('${directory.path}/deferred/${value.objectId}.bin'),
      value.bytes,
    );
  }

  void _writeAtomically(File target, Uint8List bytes) {
    final temp = File('${target.path}.tmp');
    temp.writeAsBytesSync(bytes, flush: true);
    try {
      temp.renameSync(target.path);
    } catch (_) {
      target.writeAsBytesSync(bytes, flush: true);
      if (temp.existsSync()) temp.deleteSync();
    }
  }

  static Uint8List _int64(int value) =>
      (ByteData(8)..setInt64(0, value, Endian.big)).buffer.asUint8List();

  static Uint8List _int32(int value) =>
      (ByteData(4)..setInt32(0, value, Endian.big)).buffer.asUint8List();
}

File: core-protocol/src/main/kotlin/in/meshsetu/protocol/RelayEngine.kt
package `in`.meshsetu.protocol

import `in`.meshsetu.model.EncryptedObject
import `in`.meshsetu.model.MeshEnvelope
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.file.Files
import java.nio.file.StandardCopyOption

interface RelayStore {
    fun persist(envelope: MeshEnvelope)
    fun contains(objectId: ULong): Boolean = false
    fun enqueue(value: EncryptedObject) = Unit
    fun pending(nowMs: Long): List<EncryptedObject> = emptyList()
    fun markAck(objectId: ULong, peerId: String) = Unit
    fun markDeferred(value: EncryptedObject) = Unit
}

data class RelayMetric(val kind: String, val objectId: ULong? = null, val peerId: String? = null, val value: Long? = null)

data class RelayResult(val controlFrames: List<ByteArray>, val metrics: List<RelayMetric>)

class MeshRelayEngine(
    private val siteId: String,
    private val crypto: CryptoEnvelope,
    private val store: RelayStore,
    private val clockMs: () -> Long,
    private val scheduler: OutboundScheduler = OutboundScheduler(),
    private val dedupe: RecentObjectCache = RecentObjectCache(),
    private val onPersist: (MeshEnvelope, String) -> Unit = { _, _ -> },
) {
    private data class Key(val peerId: String, val objectId: ULong)
    private val partial = mutableMapOf<Key, ReassemblyBuffer>()
    private val partialCreatedAt = mutableMapOf<Key, Long>()
    private val lastNackAt = mutableMapOf<Key, Long>()
    private val metrics = mutableListOf<RelayMetric>()
    private val listeners = mutableListOf<(MeshEnvelope, String) -> Unit>()
    private val knownObjects = mutableMapOf<ULong, EncryptedObject>()
    private val inFlight = mutableMapOf<Key, Long>()
    private val acknowledged = mutableSetOf<ULong>()

    init {
        store.pending(clockMs()).forEach { value ->
            knownObjects[value.objectId] = value
            scheduler.enqueue(value, clockMs())
        }
    }

    @Synchronized fun addPersistListener(listener: (MeshEnvelope, String) -> Unit) { listeners += listener }

    @Synchronized
    fun receive(peerId: String, encodedFrame: ByteArray): RelayResult {
        cleanupPartial(clockMs())
        val frame = runCatching { FrameCodec.decode(encodedFrame) }.getOrElse {
            metrics += RelayMetric("invalid_frame", peerId = peerId)
            return RelayResult(emptyList(), drainMetrics())
        }
        if (frame.type == FrameType.CUSTODY_ACK) {
            if (frame.payload.size != 8 || ByteBuffer.wrap(frame.payload).order(ByteOrder.BIG_ENDIAN).long.toULong() != frame.objectId) {
                metrics += RelayMetric("invalid_ack", frame.objectId, peerId)
                return RelayResult(emptyList(), drainMetrics())
            }
            if (inFlight.remove(Key(peerId, frame.objectId)) != null) {
                acknowledged += frame.objectId
                store.markAck(frame.objectId, peerId)
                metrics += RelayMetric("ack", frame.objectId, peerId)
            } else {
                metrics += RelayMetric("unexpected_ack", frame.objectId, peerId)
            }
            return RelayResult(emptyList(), drainMetrics())
        }
        if (frame.type == FrameType.NACK) {
            if (frame.payload.isEmpty()) {
                metrics += RelayMetric("invalid_nack", frame.objectId, peerId)
            } else {
                knownObjects[frame.objectId]?.let { value ->
                    if (!acknowledged.contains(frame.objectId)) {
                        scheduler.enqueue(value, clockMs())
                        metrics += RelayMetric("nack_retry", frame.objectId, peerId)
                    }
                } ?: metrics.add(RelayMetric("unknown_nack", frame.objectId, peerId))
            }
            return RelayResult(emptyList(), drainMetrics())
        }
        if (frame.type == FrameType.HELLO) {
            metrics += RelayMetric("hello", peerId = peerId)
            return RelayResult(emptyList(), drainMetrics())
        }
        if (frame.type != FrameType.DATA) return RelayResult(emptyList(), drainMetrics())
        val key = Key(peerId, frame.objectId)
        val buffer = partial[key] ?: run {
            if (partial.size >= MAX_PARTIAL_OBJECTS) {
                metrics += RelayMetric("reassembly_capacity", frame.objectId, peerId)
                return RelayResult(emptyList(), drainMetrics())
            }
            ReassemblyBuffer(frame.count.toInt()).also {
                partial[key] = it
                partialCreatedAt[key] = clockMs()
            }
        }
        runCatching { buffer.add(frame) }.onFailure {
            partial.remove(key)
            partialCreatedAt.remove(key)
            lastNackAt.remove(key)
            metrics += RelayMetric("reassembly_rejected", frame.objectId, peerId)
            return RelayResult(emptyList(), drainMetrics())
        }
        if (!buffer.complete()) return RelayResult(emptyList(), drainMetrics())
        partial.remove(key)
        partialCreatedAt.remove(key)
        lastNackAt.remove(key)
        val encrypted = EncryptedObject(frame.objectId, trafficClass(frame.priority), buffer.join(), Long.MAX_VALUE)
        val envelope = crypto.decrypt(encrypted).getOrElse {
            metrics += RelayMetric("invalid_object", frame.objectId, peerId)
            return RelayResult(emptyList(), drainMetrics())
        }
        if (envelope.siteId != siteId) {
            metrics += RelayMetric("wrong_site", envelope.objectId, peerId)
            return RelayResult(emptyList(), drainMetrics())
        }
        val now = clockMs()
        if (envelope.expiresAtMs <= now) {
            metrics += RelayMetric("expired", envelope.objectId, peerId)
            return RelayResult(emptyList(), drainMetrics())
        }
        if (store.contains(envelope.objectId) || !dedupe.markIfNew(envelope.objectId, envelope.expiresAtMs, now)) {
            metrics += RelayMetric("duplicate", envelope.objectId, peerId)
            return RelayResult(listOf(ack(frame.objectId, frame.priority)), drainMetrics())
        }
        runCatching { store.persist(envelope) }.onFailure {
            metrics += RelayMetric("persist_failed", envelope.objectId, peerId)
            return RelayResult(emptyList(), drainMetrics())
        }
        onPersist(envelope, peerId)
        listeners.forEach { it(envelope, peerId) }
        metrics += RelayMetric("object_complete", envelope.objectId, peerId, envelope.hopCount.toLong())
        if (envelope.hopCount < envelope.hopLimit) {
            enqueue(crypto.encrypt(envelope.copy(hopCount = envelope.hopCount + 1)), now)
            metrics += RelayMetric("relay_enqueued", envelope.objectId, peerId)
        }
        return RelayResult(listOf(ack(frame.objectId, frame.priority)), drainMetrics())
    }

    @Synchronized fun nextOutbound(nowMs: Long = clockMs()): EncryptedObject? = scheduler.next(nowMs)

    @Synchronized fun requeue(value: EncryptedObject, nowMs: Long = clockMs()) = enqueue(value, nowMs)

    @Synchronized fun markSent(value: EncryptedObject, peerId: String, nowMs: Long = clockMs()) {
        knownObjects[value.objectId] = value
        inFlight[Key(peerId, value.objectId)] = nowMs
    }

    @Synchronized fun retryExpired(nowMs: Long = clockMs(), timeoutMs: Long = 8_000) {
        val due = inFlight.filterValues { nowMs - it >= timeoutMs }.keys.toList()
        due.forEach { key ->
            inFlight.remove(key)
            if (key.objectId !in acknowledged) knownObjects[key.objectId]?.let { enqueue(it, nowMs) }
            metrics += RelayMetric("ack_timeout", key.objectId, key.peerId)
        }
    }

    @Synchronized fun defer(value: EncryptedObject) {
        knownObjects.remove(value.objectId)
        store.markDeferred(value)
    }

    @Synchronized fun missing(peerId: String, objectId: ULong, priority: UByte): ByteArray? {
        val buffer = partial[Key(peerId, objectId)] ?: return null
        return FrameCodec.encode(MeshFrame(FrameType.NACK, priority, 0u, objectId, 0u, 1u, buffer.missingBitmap()))
    }

    @Synchronized fun missingForPeer(peerId: String, priority: UByte = 3u, nowMs: Long = clockMs()): List<ByteArray> = partial.keys
        .filter { it.peerId == peerId && nowMs - (partialCreatedAt[it] ?: nowMs) >= NACK_DELAY_MS }
        .mapNotNull { key ->
            if (nowMs - (lastNackAt[key] ?: Long.MIN_VALUE) < NACK_DELAY_MS) return@mapNotNull null
            lastNackAt[key] = nowMs
            missing(peerId, key.objectId, priority)
        }

    @Synchronized fun submit(envelope: MeshEnvelope, nowMs: Long = clockMs()): EncryptedObject {
        val encrypted = crypto.encrypt(envelope)
        enqueue(encrypted, nowMs)
        return encrypted
    }

    private fun enqueue(value: EncryptedObject, nowMs: Long) {
        knownObjects[value.objectId] = value
        store.enqueue(value)
        scheduler.enqueue(value, nowMs)
    }

    private fun cleanupPartial(nowMs: Long) {
        val expired = partialCreatedAt.filterValues { nowMs - it >= PARTIAL_TIMEOUT_MS }.keys.toList()
        expired.forEach {
            partial.remove(it)
            partialCreatedAt.remove(it)
            lastNackAt.remove(it)
            metrics += RelayMetric("reassembly_timeout", it.objectId, it.peerId)
        }
    }

    @Synchronized fun drainMetrics(): List<RelayMetric> = metrics.toList().also { metrics.clear() }
    private fun trafficClass(priority: UByte) = when (priority.toInt()) { 1 -> `in`.meshsetu.model.TrafficClass.SOS_STRUCTURED; 2 -> `in`.meshsetu.model.TrafficClass.AUTHORITY_CONTROL; 3 -> `in`.meshsetu.model.TrafficClass.VOICE_EVIDENCE; 4 -> `in`.meshsetu.model.TrafficClass.ROOM_MESSAGE; else -> `in`.meshsetu.model.TrafficClass.TELEMETRY }
    private fun ack(objectId: ULong, priority: UByte): ByteArray = FrameCodec.encode(MeshFrame(FrameType.CUSTODY_ACK, priority, 0u, objectId, 0u, 1u, ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN).putLong(objectId.toLong()).array()))

    private companion object {
        const val MAX_PARTIAL_OBJECTS = 128
        const val PARTIAL_TIMEOUT_MS = 30_000L
        const val NACK_DELAY_MS = 2_000L
    }
}

class FileRelayStore(private val directory: File) : RelayStore {
    private val inbox = File(directory, "inbox")
    private val outbox = File(directory, "outbox")

    init { inbox.mkdirs(); outbox.mkdirs(); File(directory, "deferred").mkdirs() }

    @Synchronized override fun persist(envelope: MeshEnvelope) = writeAtomically(File(inbox, "${envelope.objectId}.bin")) { output ->
        output.write(EnvelopeCodec.encode(envelope))
    }

    @Synchronized override fun contains(objectId: ULong): Boolean = File(inbox, "${objectId}.bin").isFile

    @Synchronized override fun enqueue(value: EncryptedObject) = writeAtomically(File(outbox, "${value.objectId}.bin")) { output ->
        DataOutputStream(output).apply {
            writeLong(value.objectId.toLong())
            writeInt(value.trafficClass.rank)
            writeLong(value.expiresAtMs)
            writeLong(value.createdAtMs)
            writeInt(value.bytes.size)
            write(value.bytes)
            flush()
        }
    }

    @Synchronized override fun pending(nowMs: Long): List<EncryptedObject> = outbox.listFiles().orEmpty().mapNotNull { file ->
        runCatching {
            DataInputStream(FileInputStream(file)).use {
                val id = it.readLong().toULong()
                val rank = it.readInt()
                val traffic = `in`.meshsetu.model.TrafficClass.entries.first { value -> value.rank == rank }
                val expires = it.readLong()
                val created = it.readLong()
                val length = it.readInt()
                require(length in 1..MAX_OBJECT_BYTES)
                EncryptedObject(id, traffic, ByteArray(length).also(it::readFully), expires, created)
            }
        }.getOrNull()?.takeIf { it.expiresAtMs > nowMs }
    }

    @Synchronized override fun markAck(objectId: ULong, peerId: String) { File(outbox, "${objectId}.bin").delete() }

    @Synchronized override fun markDeferred(value: EncryptedObject) {
        File(outbox, "${value.objectId}.bin").delete()
        writeAtomically(File(directory, "deferred/${value.objectId}.bin")) { output -> output.write(value.bytes) }
    }

    private fun writeAtomically(target: File, write: (FileOutputStream) -> Unit) {
        val temp = File(target.parentFile, "${target.name}.tmp")
        FileOutputStream(temp).use { output -> write(output); output.fd.sync() }
        runCatching {
            Files.move(temp.toPath(), target.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
        }.getOrElse {
            Files.move(temp.toPath(), target.toPath(), StandardCopyOption.REPLACE_EXISTING)
        }
    }
}
